### ISSUE TITLE
Potential unhandledException attempting "republish" during shutdown

### ISSUE THREAD

**Issue by rossj** [guidesmiths/rascal#150]
Hi there, I just upgraded from 10.0.1 to 13.0.0 and I now sometimes get the below unhandled exception, despite having error handlers on the `Broker` and `Subscription`. I believe this happens when things have gone wrong elsewhere and I'm trying to abort / shut down the subscriptions and broker.

Looking at [this line](https://github.com/guidesmiths/rascal/blob/2b6cd92201cfdfcac1c544317ea99f32a857c53e/lib/amqp/Vhost.js#L336), it looks like `vhost.getConfirmChannel()` may return neither an error nor a channel if things are shutting down, which causes the below exception while attempting to do a `republish` operation on some message that's in-flight.  Perhaps `getConfirmChannel()` should return an error indicating that things are shutting down, which would then probably bubble up to the `ackOrNack` call that is attempting to do the republish operation.

```
webpack:////mount/node-services/node_modules/rascal/lib/amqp/SubscriberError.js?:88
          publisherChannel.on('error', next);
                           ^

TypeError: Cannot read property 'on' of undefined
    at Object.eval [as callback] (webpack:////mount/node_modules/rascal/lib/amqp/SubscriberError.js?:88:28)
    at eval (webpack:////mount/node_modules/rascal/node_modules/async/dist/async.mjs?:1409:22)
    at eval (webpack:////mount/node_modules/rascal/node_modules/async/dist/async.mjs?:419:16)
    at createChannel (webpack:////mount/node_modules/rascal/lib/amqp/Vhost.js?:336:30)
    at eval (webpack:////mount/node_modules/rascal/node_modules/async/dist/async.mjs?:3299:9)
    at Object.process (webpack:////mount/node_modules/rascal/node_modules/async/dist/async.mjs?:1525:17)
    at eval (webpack:////mount/node_modules/rascal/node_modules/async/dist/async.mjs?:1383:19)
    at Immediate.eval (webpack:////mount/node_modules/rascal/node_modules/async/dist/async.mjs?:171:41)
    at processImmediate (internal/timers.js:439:21)
```

I'll try to reproduce this error in some simpler test code.

**Comment by cressie176** [comment]
Hi @rossj,

Sorry about this. I agree, it's likely to be down to the `createChannel` function returning neither an error or a channel. I'll look to doing something similar to [this](https://github.com/guidesmiths/rascal/blob/master/lib/amqp/Subscription.js#L58-L61)

**Comment by cressie176** [comment]
Have pushed a fix. Now instead of crashing with the TypeError, SubscriberError.js yields an error via its callback when the channel is null or undefined. This will result in an error event being emitted from the subscription.

**Comment by cressie176** [comment]
@rossj It's a fairly simple [change](https://github.com/guidesmiths/rascal/commit/105aed73f385bf57545532b7a5f087b44999cc3a) but I'd appreciate a second pair of eyes if you don't mind.

**Comment by rossj** [comment]
Hi @cressie176, thanks for the quick update! The change looks good to me but I haven't been able to test it out yet. You said the new error will be emitted from the subscription, but it looks like it will also cause the next recovery strategy to be tried, or the error will be returned to the `ackOrNack` callback if there are no additional recovery strategies. Is this correct? I have a promise waiting on the `ackOrNack` response which I just don't want to be left hanging / unfulfilled.

**Comment by cressie176** [comment]
No, it should [short circuit](https://github.com/guidesmiths/rascal/blob/105aed73f385bf57545532b7a5f087b44999cc3a/lib/amqp/SubscriberError.js#L23) if a recovery strategy yields an error.

**Comment by rossj** [comment]
Thanks, that all sounds right to me then. I should end up in the 1st case with a rejected ackOrNack promise, which is what I was originally expecting / hoping for.

**Comment by cressie176** [comment]
Published as 13.0.2. Thanks for taking the time to investigate and report.
